### DISCUSSION TITLE
API Rules filer based on another collection

### DISCUSSION THREAD

**Discussion by srowe0091** [pocketbase/pocketbase#4710] [Q&A] [answered]
Hello, hopefully I have a simple use case here. I'm trying to integrate a "groups" relationship amongst users. The scenario is if 2 users are in the same group, they are able to see each others row in a collection if the shared flag is true.

psuedo table example

| id  | owner (relation to single user) | message | shared |
| --- | ------------------------------- | ------- | ------ |
| 1   | user1                           | lorem   | true   |
| 2   | user2                           | ipsum   | true   |
| 3   | user1                           | lorem   | false  |
| 4   | user3                           | ipsum   | true   |
| 5   | user2                           | lorem   | false  |

so by the table above, if user1 and user2 are in the same group
user1 sees [1, 2, 3]
user2 sees [1, 2, 5]

and if user1 and user3 are in the same group
user1 sees [1, 3, 4]
user3 sees [1, 4]

Group table implementations I have tried.

1. have each user be a row with just 2 ids
`group` collection
   | id | user1 (relation to single user) | user2 (relation to single user) |
   | --- | ------------------------- | ------------------------- |
   | 1 | user1 | user2 |
   | 1 | user1 | user3 |

3. each group has an array of the users that in the same group
`group2` collection
   | id | users (relation to many users) |
   | --- | ------------------------------ |
   | 1 | [user1, user2] |
   | 2 | [user1, user3] |

The API rules is where i need help with, as im not very good at this part. any help will be much appreciated!

**Comment by ganigeorgiev** [comment]
I'm not entirely sure that I understand the provided example, nor what actually is the question, but the API rule will depend on how do you plan to store the user-group relation so I'm not sure what to recommend here.

If we assume that you have the following collections:

- `groups` (`base` collection):
   - `title` (`text` field)

- `users` (`auth` collection):
   - `groups` (multiple `relation` field to the `groups` collection, containing the groups the user is part of)

- `pseudoTable` (`base` collection)
   - `owner` (single nonempty `relation` field to the `users` collection)
   - `message` (`text` field)
   - `shared` (`bool` field)

An example List API rule for the `pseudoTable` collection could be:

```js
@request.auth.id != "" && (
    // always allow access to the owner
   owner = @request.auth.id ||
   // allow access to other users if they share a group with the owner and the record has the shared flag
   (shared = true && @request.auth.groups.id ?= owner.groups.id)
)
```

If the above doesn't help, please elaborate a little more on your use case and export your actual collections structure as json from the _Admin UI > Settings > Export collections_.

**Reply by ganigeorgiev** [reply]
Alternatively, if the `user-group` relation is stored in a separate table, for example:

- `userGroup` (`base` collection)
  - `user` (single nonempty `relation` to the `users` collection)
  - `group` (single nonempty `relation` to the `groups` collection)
  - (unique index for the user-group pair to prevent duplicated records)
  
Then one possible List API rule for the `pseudoTable` collection could be:

```js
@request.auth.id != "" && (
    // always allow access to the owner
   owner = @request.auth.id ||
   // allow access to other users if they share a group with the owner and the record has the shared flag
   (shared = true && @request.auth.userGroup_via_user.group ?= owner.userGroup_via_user.group)
)
```
_(since there is no direct relation from `users` to `userGroup` we use the `_via_`, aka. [back-relation reference syntax](https://pocketbase.io/docs/working-with-relations/#back-relations))_

**Reply by srowe0091** [reply]
Sorry that i wasn't clear. The 2 collections shown at the bottom of the post was the `groups` collection that I have tried. first collection showing where 1 row was just the relation between 2 users with a column for userOne and column for userTwo. the second collection would show all the users in the group with the multiple relations.

nonetheless heres the exported collections of the implementations i have tried as well as the psuedoTable that needs the API rules

also for the record, the users auth collection does not have columns for groups.
```
[
    {
        "id": "ocwk9kmxck8i7dk",
        "name": "psuedoTable",
        "type": "base",
        "system": false,
        "schema": [
            {
                "system": false,
                "id": "pxeepzlx",
                "name": "field",
                "type": "relation",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "collectionId": "_pb_users_auth_",
                    "cascadeDelete": false,
                    "minSelect": null,
                    "maxSelect": 1,
                    "displayFields": null
                }
            },
            {
                "system": false,
                "id": "4zclyirk",
                "name": "message",
                "type": "text",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "min": null,
                    "max": null,
                    "pattern": ""
                }
            },
            {
                "system": false,
                "id": "gui9lyk7",
                "name": "shared",
                "type": "bool",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {}
            }
        ],
        "indexes": [],
        "listRule": null,
        "viewRule": null,
        "createRule": null,
        "updateRule": null,
        "deleteRule": null,
        "options": {}
    },
    {
        "id": "s62givrblwau00f",
        "name": "group2",
        "type": "base",
        "system": false,
        "schema": [
            {
                "system": false,
                "id": "hysk29b0",
                "name": "users",
                "type": "relation",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "collectionId": "_pb_users_auth_",
                    "cascadeDelete": false,
                    "minSelect": null,
                    "maxSelect": null,
                    "displayFields": null
                }
            }
        ],
        "indexes": [],
        "listRule": null,
        "viewRule": null,
        "createRule": null,
        "updateRule": null,
        "deleteRule": null,
        "options": {}
    },
    {
        "id": "sh6cqzo6vpg9sef",
        "name": "groups",
        "type": "base",
        "system": false,
        "schema": [
            {
                "system": false,
                "id": "kds5j7ld",
                "name": "userOne",
                "type": "relation",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "collectionId": "_pb_users_auth_",
                    "cascadeDelete": false,
                    "minSelect": null,
                    "maxSelect": 1,
                    "displayFields": null
                }
            },
            {
                "system": false,
                "id": "ih8jit4v",
                "name": "userTwo",
                "type": "relation",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "collectionId": "_pb_users_auth_",
                    "cascadeDelete": false,
                    "minSelect": null,
                    "maxSelect": 1,
                    "displayFields": null
                }
            }
        ],
        "indexes": [],
        "listRule": null,
        "viewRule": null,
        "createRule": null,
        "updateRule": null,
        "deleteRule": null,
        "options": {}
    }
]
```

**Reply by ganigeorgiev** [reply]
Unfortunately the above doesn't really help clarifying what is the expected outcome.

Without being specific on your desired collection structure I really don't know how to help further. As illustrated above there are multiple ways to represents the user-group relation and the API rule will depend on your specific fields. There isn't a single rule that will work for all cases, if that's what you are looking for.

For general documentation about the API rules syntax please refer to https://pocketbase.io/docs/api-rules-and-filters/#filters-syntax. Suggestions to improve the parts that are not clear are welcomed.

In any case, I'll stop engage to avoid causing further confusion, but as a last attempt, here are 2 examples `psuedoTable` API rules using the `groups` and `group2` collection:


- using `groups`
    ```js
    @request.auth.id != "" && (
        // always allow access to the owner
        owner = @request.auth.id ||
        // allow access to other users if the owner and the current auth user have a shared groups record
        (
            shared = true && (
                // either auth-owner or owner-auth match
                @request.auth.groups_via_userOne.userTwo ?= owner ||
                @request.auth.groups_via_userTwo.userOne ?= owner
            )
        )
    )
    ```

- using `group2`
    ```js
    @request.auth.id != "" && (
        // always allow access to the owner
        owner = @request.auth.id ||
        // allow access to other users if the owner is in one of the current auth groups
        (shared = true && @request.auth.group2_via_users.users.id ?= owner)
    )
    ```

**Comment by srowe0091** [comment, marked as answer]
I figured it out!! thats exciting lol. i was able to do it using the back-relations with the `group2` collection setup. I still don't understand back relations but this at least works :D

this is the API rule that works how i want it to.
`owner = @request.auth.id || (@request.auth.group2_via_users.users ?~ owner.id && group = True)`

Thank you for the help Gani and giving me the back-relations idea

**Reply by ganigeorgiev** [reply]
It seems that we commented at the same time. See the above answer https://github.com/pocketbase/pocketbase/discussions/4710#discussioncomment-9029443/. Your rule is fine but if you want an exact `id` match prefer the equal operator (`~` translates to a `LIKE` expression, aka. it will loosely search for the owner's id in the serialized json array `'["id1", "id2"]' LIKE '%{:ownerId}%'`).
